### PR TITLE
CommonTestHelper.psm1: Prevent multiple calls to git clone or git pull of the same module in the same build

### PR DESCRIPTION
**Pull request by mhendric**
#### Pull Request (PR) description
Updates test helper Enter-DscResourceTestEnvironment to only clone or pull DscResource.Tests a maximum of one time per build. Currently it will clone or pull every time this function is hit. This doesn't directly address #505, but makes it far less likely to occur, as git clone will only be called once per build, instead of once per test file.

#### This Pull Request (PR) fixes the following issues
None

#### Task list
- [x] Added an entry under the Unreleased section of the change log in the README.md.
      Entry should say what was changed, and how that affects users (if applicable).
- [ ] Resource documentation added/updated in README.md.
- [ ] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [ ] Comment-based help added/updated.
- [ ] Localization strings added/updated in all localization files as appropriate.
- [ ] Examples appropriately added/updated.
- [ ] Unit tests added/updated. See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [ ] Integration tests added/updated (where possible). See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [x] New/changed code adheres to [DSC Resource Style Guidelines](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md) and [Best Practices](https://github.com/PowerShell/DscResources/blob/master/BestPractices.md).

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/516)

### CODE REVIEW

**Comment by mhendric** [comment]
Requesting a code review on this one.

**Comment by PlagueHO** [comment]
On me! Getting it done!

**Comment by mhendric** [comment]
@PlagueHO , are you suggesting we still do either a git clone or a git pull, and that we only do the git pull if the directory is older than a certain time? Or are you suggesting we only do the git clone, but do some check on the date there as well?

And I think we need to reconcile this with @johlju 's suggestion from #505 . Do we want to go the direction of just using the test templates (and if we want to add retries and or stale checking, we address that in the template directly)? Or do we want to keep the helper function for now? I think I'd prefer the first option, but I'm not opposed to the second.

**Comment by johlju** [comment]
I think adding resilience to the test templates so it can be used by more repos. 

When we refactor the tests we will (should) add the correct template, so change that to be better at cloning the repo. 

But the problem we seen is rare. For the last two years I just seen it. So question is, should we bother with it? It is easy enough to kick off the tests again. 
If AppVeyor is having problem, more than a minute long then looping and testing again will just take time of the 60 minute window and the job will fail anyway (for a contributor account). 
But surly, I agree if there are glitches that would work if testing a couple of seconds later then I agree, we could build resilience.

**Comment by PlagueHO** [comment]
@mhendric, @johlju - I think we have two separate issues:
1. Performance - attempting a Git Pull every time a test file runs (~25 in xPSDesiredStateConfiguration) slows things down.
2. Resilience - when there is an issue accessing GH. As with @johlju, I've only begun seeing this recently (last month or so).

@mhendric - my suggestion was primarily to address 1 (but could reduce problems with 2), without introducing the problem of the DSCResources.Tests getting stale on a users local repo.

The logic might look like this:
1. If DSCResources.Tests folder does not exist then `Git Clone` it.
2. If DSCResources.Tests folder exists but FETCH_HEAD does not exist and ORIG_HEAD exists and is older than 60 minutes then `Git Pull` it.
3. If DSCResources.Tests folder does exist and FETCH_HEAD exists and is older than 60 minutes, then `Git Pull` it.
4. Otherwise don't `Git Pull` or `Git Clone`

It would be ideal if this could be proved out and then added to Template and/or migrated to existing repos. I'd be keen to improve all the resources I maintain with this as it would improve the test time a little bit.

**Comment by johlju** [comment]
I like the logic that @PlagueHO suggest. I also think we should focus on performance. 

I wouldn't say no to more resilience either, but just so we don't put in a lot of work to fix something that happens very rarely.
I have pushed like ~100 builds this weekend, and have not seen any problem with the test framework, AppVeyor, or GitHub.

**Comment by mhendric** [comment]
I'm kind of questioning whether making sure DscResoures.Tests is up to date (and not just present) should be in scope for this module. We know the AppVeyor build process will always grab the latest clone since it's not present to begin with, and we also know the Unit and Integration tests will trigger a download of the module if it's not present, even if not run in AppVeyor. I personally think we should leave it up to consumers to ensure that their copy is up to date, in the event that they have a long lived clone of the repo. Keeping it up to date is definitely achievable, but I'm not sure it's worth the added effort.

Also, it sounds like we still need to determine whether we should keep the helper function that currently does the git clone/pull, or whether we should just add the git clone operation from the test templates directly to each file. I do like the idea of keeping common code in helper functions, and having as little duplicate code as possible spread throughout the test files. Perhaps the DSC test templates should be modified to include a test helper file, which will include the functionality to git clone/pull.

What do you guys think?

**Comment by mhendric** [comment]
@PlagueHO , regarding your comment of not using Set/Pop-Location. I'm not sure we can do git pull without changing to the directory of the repo (I don't see any directory specific parameters for it). Can you think of a way to avoid changing directory?

Also, I'm still wondering if keeping DscResource.Tests up to date shouldn't be handled by the caller, or by the module itself What do you think of putting the check to see if it's up to date in Initialize-TestEnvironment, since that's called in the DSC test templates?

**Comment by PlagueHO** [comment]
@mhendric - you're right. It doesn't look like Git supports that. I thought it did (maybe just git clone).

Thinking about your proposal around putting the check in Initialize-TestEnvironment. That is quite interesting. Hadn't thought of that approach. That could work, but the user would have to first update to the version that had the newer Initialize-TestEnvironment -that is not a huge problem though. @johlju - what do you reckon?

**Comment by mhendric** [comment]
Hey @PlagueHO and @johlju , just following up to see what you guys want to do with this one. Also, just wanted to throw out one other option for this. We could potentially just leave the existing code mostly as is, and add a (dreaded) global variable to track whether the git pull has been done, so that we can be sure to only do it once per job. That might be easier than checking modified dates on various GIT files. I'd still prefer to not do the latest version check at all in this module, but I'm willing to compromise.

**Comment by PlagueHO** [comment]
I've had a few conversations around this and the general feeling is that we should not be pulling direct from Git - rather using a packaged artefact. Of course being able to pull from Git if you wish would be important for debugging and testing this repo. I've spent a lot of time at work telling teams not to do what we are doing here - so it feels a bit wrong that we are doing it. So maybe we are best investigating in how we might package this?

**Comment by mhendric** [comment]
Well I think we have a couple issues here.

1. One is that we have an issue where the 'git pull' code is being called by every single test in the job, which adds time to the overall test run, and if AppVeyor is unable to connect to Git at any time, the test will fail. 

2. The other is that we need a better way to make sure helper Modules (like DSCResource.Tests, and possibly Pester), are the latest and greatest when tests are executed.

I'd personally like to de-couple these and treat 1 as a short term problem and get that fixed ASAP to make our tests more resilient. Then deal with 2 separately. I don't personally have much experience with packaging, so I'd probably defer to someone else for that.

What do you think?

**Comment by mhendric** [comment]
@PlagueHO , I've pushed an alternate option that will just address number 1 from above. Let me know what you think.

**Comment by mhendric** [comment]
Ugggg… a Global isn't going to work here since we're in Strict Mode (at least through AppVeyor...), which prevents us from checking if the variable is null. If we want to track this with a global we'd have to temporarily turn off strict mode.

**Comment by tmeckel** [comment]
@mhendric how about using `Test-Path variable:\myvarname` to check if the variable exists? That's the way I check if a variable exists before I try to access it, when the script is running in `Strict Mode`.

**Comment by mhendric** [comment]
Looks like that may just do the trick. Thanks for the tip @tmeckel !

**Comment by johlju** [comment]
I wonder, couldn't we just clone this if the DscResource.Tests does not exist, and if the DscResource.Tests folder is older than, say, a week the folder is removed and re-cloned?

**Comment by tmeckel** [comment]
As I commented here https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/556#issuecomment-464454524 my overall idea was to replace pulling DscResource.Tests with defining DscResource.Tests as a submodule to the repository and adding a merge hook that keeps the submodule uptodate whenever a (merge) pull request is done.

**Comment by johlju** [comment]
@tmeckel I run tests a lot without first adding a commit to the local working branch. So I think this need to work with users running `Invoke-Pester` locally outside, and for those that never use git. 🤔

**Comment by tmeckel** [comment]
@johlju the merge hook is also run when you clone the repo initially. And I'm unsure what you mean by 
 "... and for those that never use git. " And yes perhaps only having a merge hook, which runs on a `pull`, isn't enough because this hook doesn't run on `clone`, but because a clone is a `fetch` with a `checkout` we can utilize the `post-checkout` as well to have an update everytime you a checkout.

https://www.digitalocean.com/community/tutorials/how-to-use-git-hooks-to-automate-development-and-deployment-tasks

**Comment by johlju** [comment]
> "... and for those that never use git. "

@tmeckel I mean that we package the tests in the release package, so if a user want to test the resources that they install through `Install-Module` they are probably not using git to commit or anything else. Today a requirement is that git is installed to run the tests though. But that is another issue that was touched here https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/516#issuecomment-464144301 to get this another way.

**Comment by johlju** [comment]
@tmeckel I do like your idea, but there are scenarios today that it does not support I think. 🤔

**Comment by tmeckel** [comment]
> > "... and for those that never use git. "
> 
> @tmeckel I mean that we package the tests in the release package, so if a user want to test the resources that they install through `Install-Module` they are probably not using git to commit or anything else. Today a requirement is that git is installed to run the tests though. But that is another issue that was touched here [#516 (comment)](https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/516#issuecomment-464144301) to get this another way.

Okay understood ... when I assume that the packaging is done when the resource is pushed to the PowershellGallery. Right?  And for sure I guess the packing is done via a script itsself. Right? 

But when I look at the current nuget package there's no DSCResource.Tests folder in there. So people are still forced to install git on a machine where they want to run the tests and that fits to your statement  `... git is installed to run the tests though.` and it explains the changes @mhendric tries to add with this PR.

Therefore if we add the the DSCResource.Tests as submodule to xPSDesiredStateConfiguration and we add the mentioned hooks, the DSCResource.Tests willl checked out and updated when a clone is done in preparation for a release package. Thus the NuGet package will contain the DSCResource.Tests subfolder with all the required files for running the tests "offline" without the requirement to have git installed because the test alway try to pull DSCResource.Tests from GitHub. Another benefit here is that DSCResource.Tests is statically shipped i.e. the NuGet package of a DSC Module always contains the version of DSCResource.Tests with it has been tested against. With the current git pull approach you might break tests of a DSC Module because of changes in DSCResource.Tests.

In addtion to that, a developer who's working on the module has always the current version DSCResource.Tests at hand because during a pull DSCResource.Tests is updated via the merge-hook. In that case we can neglect issues with changes to DSCResource.Tests because development is done on the 'dev' branch anyway and should 'fix' issues with a changed implementation of DSCResource.Tests anyway.

Do I miss something?

**Comment by mhendric** [comment]
Hey guys, I just want to throw out that we are still semi-regularly hitting the issue described in #505 , which forces the owner of a PR to submit a new commit just to retry the tests and hope that one or more of the 50 or so git calls that the current function makes don't fail. If possible, I'd really like to get a short term fix in for this particular issue ASAP to make that scenario less likely to happen (i.e. we'll only do a git call once instead of 50 or so times). I'm fine with checking the directory age vs looking at a global variable if that's what everyone thinks is the better approach. Hopefully we can all agree on what the best approach is in the short term, we can get that checked in, and then we can spin off an issue to work on a long term fix.

The above being said, @PlagueHO , @johlju , @tmeckel , what do you think is the best short term fix for this? Should we use a global variable to track whether we've tried to update the folder for that job? Or should we check directory or file age?

**Comment by tmeckel** [comment]
@mhendric why not adding a detection if the test is running on Appveyor. If so a test if DSCResource.Tests already exist should be sufficient. In any other case leave the current implementation as is.

**Comment by tmeckel** [comment]
https://help.appveyor.com/discussions/questions/271-how-to-detect-that-my-tests-are-running-under-appveyor-ci

**Comment by johlju** [comment]
@mhendric take the options an weight them, and do the one you think is the best for now, we can always change it later if need be. 🙂

**Comment by mhendric** [comment]
Ok. I changed this to perform the git pull if the newest file in the DSCResource.Tests folder is over 60 minutes old. That check will only be performed once per job. I did not do any AppVeyor specific tests, as I don't think we want git pull getting called 50-ish times even if we aren't running under AppVeyor. Let me know what you guys think. Thanks.

**Comment by tmeckel** [comment]
Super as a quick solution! But for the problem, which has existed for a long time now and affects many (all?) DSC modules, a real solution should be worked out though.

**Comment by PlagueHO** [comment]
I like this as a solution for now @mhendric. I think we should implement this and try it for a bit and see how it goes. I reckon we want to not convert all other repos over until we've tested this out.

**Comment by mhendric** [comment]
Hey guys. Perhaps we should take the long term discussion for this over to #505 (or potentially a new issue altogether). @tmeckel , your proposal sounds good, although I'm having a hard time visualizing exactly how it would be implemented and if it would in fact solve the problem of DSCResource.Tests getting outdated. Would you be able to create a proof of concept for this (perhaps after #507 gets merged)?

And also, just to call out, I don't think the problem I'm trying to address with this particular PR (#516) exists in other modules. This is the only module that I'm aware of that tries a git pull operation for every Integration and Unit test. That said, I think it's a good idea to come up with a better way for modules to obtain and keep up to date any dependent modules.

@PlagueHO , if you're OK with this short term fix, can you do an official review again so we can merge and close this one out?

Thanks.

**Comment by tmeckel** [comment]
Hi @mehendric using git hooks was a nice idea of mine but what I didn’t take into account was the fact that hooks aren’t part of a repository because they’re stored inside the `.git` directory. Thus hooks aren‘t versioned and not directly distributable. 

When you think about this a bit it’s only logical that hooks aren’t versioned because that might imply security issues, since hooks can contain arbitrary commands that are executed on the computer to which you clone a repository. So to have hooks versioned you have to store the definitions in a separate directory inside the repo. But that introduces the next issue because Git (normally) executes hooks defined in the `.git\hooks` directory only. 

Starting with Git 2.9 you can define the location of the hooks directory via the `core.hooksPath` configuration settings. For older Git versions the hook definitions must be copied from the hooks directory inside the repository into the `.git\hooks` directory (symlinking is another option as well). 

Both solutions require manual intervention of the developer and thus the option using submodules isn’t any better compared to the current solution. Well is has some advantages which I described already but when it comes down to a completely automated process it has it obstacles too. Repositories which contain code that has to be built in some way integrate the hook configuration steps inside a makefile or in an npm (package.config) configuration for example. 

With Powershell we don’t need such a build infrastructure and thus you’d have to explain to a developer that e.g. a special init script has to run to configure the cloned repository for starters which people often forget though. However in a build pipeline this could be achieved easily. Even though referencing DSCResource.Tests as a submodule would give us some advantages a complete automated process is hard to be done. Whereas I must say that you only have to configure the hooks ONLY ONCE when you've freshly cloned the repository. So it should be feasible for a potential developer to do this. 

Another option besides having a section about how to configure the hooks for a DSC repository in the `README.md` would be that the `Enter-DscResourceTestEnvironment` method could check and configure this once the tests are run. 

**Long story short:** submodules aren’t the silver bullet here either.

**Comment by johlju** [comment]
@tmeckel thanks for the rundown on the hooks!

**Comment by mhendric** [comment]
Can someone please give this one a new, official review? I'd really like to either get this checked in, or just close it out if no one else thinks this change should go in. I don't want to unilaterally make the decision to check this in, and skip the official review process. Thanks.

**Comment by PlagueHO** [comment]
@mhendric - will get into this one tonight. Sorry about delay!

**Comment by mhendric** [comment]
Thanks @PlagueHO . This issue got me again today :( https://ci.appveyor.com/project/mhendric/xpsdesiredstateconfiguration/build/job/wtlpuu84rp3olwm9

**Comment by gaelcolas** [comment]
Wow. Just catching up on this one.

I think that:

1. As @PlagueHO said, the problem is the reliance of git instead of versioned immutable artefacts. That needs to be addressed but it's out of scope here (should move to [DscResource.Template](https://github.com/PowerShell/DscResource.Template), so I'll open an RFC this week).
2. Addressing the problem for **this repo** is the priority until we get something we can generalise, as long as we're not compromising quality. This PR works as a temporary & isolated workaround.
3. I'm ok with the changes here. @PlagueHO  What do you think? Can you close the review?

**Comment by mhendric** [comment]
@PlagueHO , to be honest, I'm not sure that a magic file is much better than a Global variable. The magic file would pretty much just provide a single location for different pieces of code to check if it's set, and then update if it's not. Sounds a lot like a Global variable to me. The main difference I see is that magic files are not explicitly banned by PSScriptAnalyzer. Other than that, it seems they would serve the same purpose. What do you think?

**Comment by mhendric** [comment]
Also, re: the array, I originally tried to update the git clone statement to match the DSC Test Templates in my first commit. Is that what you had in mind? I can bring that change back if so.

https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/516/commits/70e13a12ff824a1b460ba0fa6e31fc9d309cd648

**Comment by johlju** [comment]
A global variable will work for this particular repo, and can be done to get this PR through. 

But in the long run, if two repos are using the same global variable, the if a user clones one repo it will automatically update, but if the user clones another repo in the same session and runs a test there, the global is already set, so no clone of the test repo will happen. We could of course change the variable name, but then it wouldn’t be as easily reusable. 
Using the magic file, or similar, would work for each repo.

**Comment by mhendric** [comment]
Makes sense. Magic file would be per repo, but Global could potentially span multiple repo's.

**Comment by PlagueHO** [comment]
Yep @johlju covered it. That was what I was trying to suggest :grin:

**Comment by mhendric** [comment]
Hey @PlagueHO and @johlju . My intention with the Global variable was to avoid repeated calls of 'dir -recurse' to the DSCResource.Tests folder to find the newest file, which may add some overhead if the folder gets large. If we don't want to go with a Global though, I think I'd opt to just not do the magic file either, let this 'dir -recurse' happen each time, and take the perf hit.

I'm also still open to one of my first suggestions, which is just getting rid of the DSCResource.Tests "refresh" behavior that only this module implements, and just doing a single Git clone if the folder doesn't exist like every other module that follows the DSC Test Templates. If we're looking for clean solutions, that's the cleanest.

**Comment by johlju** [comment]
I can agree it’s better it follows the same method that the others does, and then we can “fix all” repos in a joint effort through the DscResources.Test and/or DscResource.Template repo.

**Comment by mhendric** [comment]
As a reminder, this was my original commit, which attempted to get this module to more closely align with the test templates, and to not handle refreshing DSCResource.Tests:

https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/516/commits/70e13a12ff824a1b460ba0fa6e31fc9d309cd648

**Comment by johlju** [comment]
I think reverting to that commit and then take it from there to get this PR merged. 
But Daniel and Mike, I think you guys as the community maintainers get to decide this. 
@PlagueHO what is your thought?

**Comment by mhendric** [comment]
@PlagueHO , here are my preferred options in order. Let me know which one you think I should pursue:

1. Remove the "refresh" functionality entirely and update function to only do a git clone if DSCResource.Tests doesn't exist (as was done here: https://github.com/PowerShell/xPSDesiredStateConfiguration/commit/70e13a12ff824a1b460ba0fa6e31fc9d309cd648)

2. Modify the existing PR to not use globals or magic file, but do everything else the same (and take the perf hit of repeated 'dir -recurse' calls)

3. Someone else submits a PR to fix the 'refresh' behavior and I close this one out :)

**Comment by PlagueHO** [comment]
Hi @mhendric - sorry about being a bit slow to respond this week. Day job has has to occupy all my time. Will have some time to catch up this weekend.

I know this one has been hanging around a while. The reason I'm trying to make sure that we end up with the most optimal solution  is that is like to also implement this in all repos I maintain. Anyway, I'm happy to do 3 and then we can figure out if it has other potential problems? I'm not completely against 1 though if we move the whole DSCResource.Tests into a PSgallery package.

**Comment by mhendric** [comment]
Alright closing this one out. @PlagueHO I've assigned you #505 so you can pick this one up. Thanks.
